fml/dart/dart_converter.h

```diff
@@ -0,0 +1,118 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef FLUTTER_FML_DART_DART_CONVERTER_H_
+#define FLUTTER_FML_DART_DART_CONVERTER_H_
+
+#include <memory>
+#include <vector>
+
+#include "flutter/fml/mapping.h"
+#include "third_party/dart/runtime/include/dart_api.h"
+#include "third_party/tonic/converter/dart_converter.h"
+
+namespace tonic {
+
+using DartConverterMapping = std::unique_ptr<fml::Mapping>;
+
+template <>
+struct DartConverter<DartConverterMapping> {
+  static Dart_Handle ToDart(const DartConverterMapping& val) {
+    if (!val) {
+      return Dart_Null();
+    }
+
+    auto dart_list_handle = Dart_NewListOf(Dart_CoreType_Int, val->GetSize());
+
+    if (Dart_IsError(dart_list_handle)) {
+      FML_LOG(ERROR) << "Error while attempting to allocate a list: "
+                     << Dart_GetError(dart_list_handle);
+      return dart_list_handle;
+    }
+
+    if (val->GetSize() == 0) {
+      // Nothing to copy. Just return the zero sized list.
+      return dart_list_handle;
+    }
+
+    auto result = Dart_ListSetAsBytes(dart_list_handle,   // list
+                                      0,                  // offset
+                                      val->GetMapping(),  // native array,
+                                      val->GetSize()      // length
+    );
+
+    if (Dart_IsError(result)) {
+      FML_LOG(ERROR) << "Error while attempting to create a Dart list: "
+                     << Dart_GetError(result);
+      return result;
+    }
+
+    return dart_list_handle;
+  }
+
+  static void SetReturnValue(Dart_NativeArguments args,
+                             const DartConverterMapping& val) {
+    Dart_SetReturnValue(args, ToDart(val));
+  }
+
+  static DartConverterMapping FromDart(Dart_Handle dart_list) {
+    if (Dart_IsNull(dart_list)) {
+      return nullptr;
+    }
+
+    if (Dart_IsError(dart_list)) {
+      FML_LOG(ERROR) << "Cannot convert an error handle to a list: "
+                     << Dart_GetError(dart_list);
+      return nullptr;
+    }
+
+    if (!Dart_IsList(dart_list)) {
+      FML_LOG(ERROR) << "Dart handle was not a list.";
+      return nullptr;
+    }
+
+    intptr_t length = 0;
+    auto handle = Dart_ListLength(dart_list, &length);
+
+    if (Dart_IsError(handle)) {
+      FML_LOG(ERROR) << "Could not get the length of the Dart list: "
+                     << Dart_GetError(handle);
+      return nullptr;
+    }
+
+    auto mapping_buffer = ::malloc(length);
```

Comment:
For a zero-length list this can return null, in which case we'll handle as an OOM below.

Reply:
Done. Good catch. Returns a valid zero sized mapping now.